Fix the REST API controllers registration

Konamiman opened the pull request:
### Changes proposed in this Pull Request:

The code in `Server::register_rest_routes` was mistakenly assuming that any controller class name not starting with `WC_REST_` was handled by the container, but other plugins register their own controller classes that use a different naming convention (for example: WooCommerce Gift Cards).

Now the container's `has` method is used instead to check if the controller class is handled by the container. As an added improvement, for classes outside of the container `LegacyProxy::get_instance_of` is used to get an instance of the class instead of just `new` (so classes with a static `instance` property or with a `load` method will be handled as expected).

<!-- Begin testing instructions -->

1. Install [the WooCommerce Gift Cards extension](https://woo.com/products/gift-cards/).
2. Create a simple product ticking the "Gift card" box, specify any amount, save.
3. Go to the shop and purchase the gift card. Make sure that the order completes.
4. Hit the `/wp-json/wc/v2/gift-cards` REST API endpoint and see how the details of the purchased gift card are returned.
5. Hit the endpoint for a core controller class that is handled by the container (e.g. `/wp-json/wc/v3/orders/<order id>/receipt`) and another one for a class that isn't (e.g. `/wp-json/wc/v3/products`), ensure that they are still working as expected.

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

#### Comment <!-- If the changes in this pull request don't warrant a changelog entry, you can alternatively supply a comment here. Note that comments are only accepted with a significance of "Patch" -->

</details>

Konamiman committed:
Fix the REST API controllers registration

The code was mistakenly assuming that any controller class name not
starting with WC_REST_ was handled by the container, but other
plugins may register their own controller classes that use
a different naming convention.

Konamiman committed:
Add changelog file

lsinger reviewed (approved):
This looks good and tests well. Thank you for the quick fix, @Konamiman! :shipit:

Aljullu committed:
Prevent script caching by using the WC core version to hash Interactivity API assets instead of old WC Blocks version (#44806)

* Prevent script caching by using the WC core version to hash Interactivity API assets instead of old WC Blocks version

* Add changelog entry

WunderBart committed:
[E2E] Stabilize Product Collection "on sale" tests (#44796)

adimoldovan committed:
Force mariadb docker image tag (#44841)

ilyasfoo committed:
Wrangle MarketingRecommendations as RemoteSpecsEngine (#44828)

* Fix potential recursion error, add MarketingRecommendations to extend RemoteSpecsEngine, add error handling, add tests

* Changelog

* Added spec evaluation and return default when empty

* Update plugins/woocommerce/src/Admin/Features/MarketingRecommendations/Init.php

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

---------

Co-authored-by: Chi-Hsuan Huang <chihsuan.tw@gmail.com>

tomalec committed:
Improve rendering of Order Attribution inputs (#44333)

- Expose public method to render OA input fields 
- Render OA inputs on `woocommerce_checkout_after_customer_details` 
   in case the checkout form does not contain `woocommerce_checkout_shipping`, `woocommerce_after_order_notes`.

Addresses part of #44159

Co-authored-by: Justin Palmer <228780+layoutd@users.noreply.github.com>
Co-authored-by: github-actions <github-actions@github.com>

corsonr committed:
Fetch & display marketplace promotions (#44840)

* Fetch & display marketplace promotions

* add changelog entry

* Pass promos via inline JS, demove unused JS file, rely on transient only

* Return early if not on expected page

* Compare dates in UTC

* remove unused global userLocale

* Small tweaks.

* TypeScript linter errors.

* fix TS lint

---------

Co-authored-by: Remi Corson <1649788+corsonr@users.noreply.github.com>
Co-authored-by: And Finally <andfinally@users.noreply.github.com>

rrennick committed:
add missing community reviewer folders (#44819)

Co-authored-by: Ron Rennick <ronald.rennick@automattic.com>

adimoldovan committed:
Fix pnpm postinstall temp workaround (#44853)

nigeljamesstevenson committed:
Update release-code-freeze.yml

andfinally committed:
Promotions: changed transient name (#44857)

* Changed transient name.

* Changelog.

nigeljamesstevenson committed:
Update release-code-freeze.yml

Maikuolan committed:
Fix typos. (#40899)

* Fix typos.

* Add changefile(s) from automation for the following project(s): woocommerce

---------

Co-authored-by: github-actions <github-actions@github.com>
